File: algorithm/leetcodeObjectiveC/3Sum-closest.m
// 16. 3Sum Closest
// Difficulty: Medium
// Given an array S of n integers, find three integers in S such that the sum is closest to a given number, target.
// Return the sum of the three integers. You may assume that each input would have exactly one solution.

// For example, given array S = {-1 2 1 -4}, and target = 1.

// The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).

// Time:  O(n^2)
// Space: O(1)

#import <Foundation/Foundation.h>

NSInteger threeSumClosest(NSArray* nums, int target) {
  NSInteger ans = INT_MAX;
  NSInteger minDiff = INT_MAX;
  
  NSArray* sorted = [nums sortedArrayUsingComparator:^NSComparisonResult(id obj1, id obj2) {
    return [obj1 compare:obj2];
  }];
  
  for (NSInteger i = 0; i < [sorted count] - 2; ++i) {
    for (NSInteger j = i + 1, k = [sorted count] - 1; j < k;) {
      NSInteger sum = [sorted[i] integerValue] + [sorted[j] integerValue] + [sorted[k] integerValue];
      
      if (sum > target) {
        --k;
      } else if (sum < target) {
        ++j;
      } else {
        return target;
      }
      
      if (labs(sum - target) < minDiff) {
        minDiff = labs(sum - target);
        ans = sum;
      }
    }
  }
  return ans;
}

File: data-structure-and-algorithm/javascript/leetcode/3-sum-closest.js
/**
Given an array nums of n integers and an integer target, find three integers in nums such that the sum is closest to target. Return the sum of the three integers. You may assume that each input would have exactly one solution.

 

Example 1:

Input: nums = [-1,2,1,-4], target = 1
Output: 2
Explanation: The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).
 

Constraints:

3 <= nums.length <= 10^3
-10^3 <= nums[i] <= 10^3
-10^4 <= target <= 10^4
*/

/**
 * 
 * @param { number[] } nums
 * @param { number } target
 * @return { number } 
 */
function threeSumClosest(nums, target) {
    let result = NaN;
    const diff = Number.MIN_VALUE;

    nums.sort((a, b) => a - b);

    for (const i = 0; i < nums.length - 2; ++i) {
        for (const left = i + 1, right = nums.length - 1; left < right;) {
            const sum = nums[i] + nums[left] + nums[right];
            const currDiff = Math.abs(sum - target);
            if (currDiff < diff) {
                diff = currDiff;
                result = sum;
            }
            if (sum < target) {
                ++left;
            } else if (sum > target) {
                --right;
            } else {
                return result;
            }
        }
    }

    return result;
}